### DISCUSSION TITLE
Compile error with include

### DISCUSSION THREAD

**Discussion by blkdragn77** [nlohmann/json#2424] [Q&A] [answered]
I have a small test app in Visual Studio 2017 that grabs JSON from an API.  I cloned this project and added the include dir   to the project properties.  When adding the #include <nlohmann/json.hpp> (and no other code) I am getting the following error when attempting to compile.

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2672	'std::declval': no matching overloaded function found	test-getapi	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.10.25017\include\type_traits	184	
Error	C2770	invalid explicit template argument(s) for 'add_rvalue_reference<_Ty>::type std::declval(void) noexcept'	test-getapi	C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\VC\Tools\MSVC\14.10.25017\include\type_traits	184	

I've searched through this project and on Google, but haven't been able to find anything as of yet.  To isolate the include, I created a new cpp project that only contains the following:

#include <nlohmann/json.hpp>

int main()
{
	return 0;
}

And am getting the same error as above.  I am assuming I am doing something wrong.  I did c/cpp dev work for over 10 years, but haven't touched it for about 5 years now, so I am a bit rusty...and I never worked in Visual Studio previously as I was always working with vi and command line compilers on a Linux platform.  So, if someone wouldn't mind pointing out where I am going wrong I would appreciate it.  :(

**Comment by calben** [comment, marked as answer]
Here are the steps I take:

1. Open Visual Studio
1. Use the Create a new project dialogue
1. Select "console app"
1. This gives me a project with an `*.sln` file and a `*.cpp` main file
1. I copy the single include directory to a new directory `ThirdParty/nlohmann/json.hpp` in the same directory as the `vcxproj` file
1. I add the `ThirdParty` directory to my includes with `$(SolutionDir)ThirdParty`
1. I can then `#include "ThirdParty/nlohmann/json.hpp`

There are definitely potential issues with your project, but maybe try going through those steps so we can double check there isn't an issue with your system setup.

You may also want to look into using `vcpkg` with `cmake` in Visual Studio 2019.   
It takes a little learning, but once you set it up and get used to it your dependency management experience is a little more like what you get from tools like `pip` and `npm`.
Having said that, it may not be an option in your project, and we'll get your existing include problem solved.

**Reply by blkdragn77** [reply]
Ok.  I just gave that a try with the single include in the ThirdParty directory.  Ran a debug build command on both x86 and x64.  The same error appeared as in my original message.  Empty main function as before.

I have looked at vcpkg and some other systems like conan.  Honestly, I am probably going to move away from developing in Windows.  I have a poweredge server that I am going to repurpose and make it my dev environment here at home.  Go back to what I know and stop trying to teach this old dog new tricks.  :)

Thank you for your assistance.  This obviously isn't a quick fix and I don't want to waste any more of my time (or anyone else's for that matter).

**Reply by blkdragn77** [reply]
Thanks again for your input @calben 
Got setup with a Ubuntu dev server and this lib works perfectly as expected.
